src/header/common/origin.rs

```diff
@@ -0,0 +1,125 @@
+/// This header is often used to inform recipients of the security context of where the request was initiated.
+///
+///
+/// Following the spec, https://fetch.spec.whatwg.org/#origin-header, the value of this header is composed of
+/// a String (scheme), header::Host (host/port)
+///
+/// # Examples
+/// ```
+/// use hyper::header::{Headers, Origin, Host};
+///
+/// let mut headers = Headers::new();
+/// headers.set(
+///     Origin{
+///           scheme: "http".to_owned(),
+///           host: Host{hostname: "hyper.rs".to_owned(), port: None}
+///    }
+/// );
+/// ```
+/// ```
+/// use hyper::header::{Headers, Origin, Host};
+///
+/// let mut headers = Headers::new();
+/// headers.set(
+///     Origin{
+///           scheme: "https".to_owned(),
+///           host: Host{hostname: "wikipedia.org".to_owned(), port: Some(443)}
+///     }
+/// );
+/// ```
+
+#[derive(Clone, Debug)]
+pub struct Origin {
+    /// The scheme, such as http or https
+    pub scheme: String,
+    /// The host, such as Host{hostname: "hyper.rs".to_owned(), port: None}
+    pub host: Host,
+}
+
+impl Header for Origin {
+    fn header_name() -> &'static str {
+        static NAME: &'static str = "Origin";
+        NAME
+    }
+
+    fn parse_header(raw: &[Vec<u8>]) -> ::Result<Origin> {
+        from_one_raw_str(raw).and_then(|s: String| {
+            let url = Url::parse(&s).unwrap(); // unwrap so we can let url::ParseError Propogate
+            let (host_str,
+                 scheme,
+                 port) = (
+                url.host_str(),
+                url.scheme(),
+                url.port_or_known_default()
+            );
+
+            let hostname = match host_str {
+                None => return Err(::Error::Header),
+                Some(hostname) => hostname.to_owned()
+            };
+            let host = Host{
+                hostname: hostname,
+                port: port
+            };
+
+            Ok(Origin{
+                scheme: scheme.to_owned(),
+                host: host
+            })
+        })
+    }
+}
+
+
+impl HeaderFormat for Origin {
+    fn fmt_header(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self.host.port {
+            None => write!(f, "{}://{}", self.scheme, self.host.hostname),
+            Some(port) => write!(f, "{}://{}:{}", self.scheme, self.host.hostname, port),
+        }
+    }
+}
+
+impl PartialEq for Origin {
+    fn eq(&self, other: &Origin) -> bool {
+        self.scheme == other.scheme && self.host == other.host
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::Origin;
+    use header::{Header, Host};
+
+    #[test]
+    fn test_origin() {
+        let origin = Header::parse_header([b"http://foo.com".to_vec()].as_ref());
+        assert_eq!(origin.ok(), Some(Origin {
+            scheme: "http".to_owned(),
+            host: Host {
+                hostname: "foo.com".to_owned(),
+                port: Some(80),
```

Comment:
This should probably be `port: None`, since the raw header didn't actually include the port. Otherwise, once writing this header again, it will appear as if the header were really `http://foo.com:80`.